README.rst
=======================================================
Reproducible and Reusable Data Analysis Workflow Server
=======================================================

.. image:: https://img.shields.io/pypi/pyversions/flowserv-core.svg
    :target: https://pypi.org/pypi/flowserv-core

.. image:: https://badge.fury.io/py/flowserv-core.svg
    :target: https://badge.fury.io/py/flowserv-core

.. image:: https://img.shields.io/badge/License-MIT-yellow.svg
    :target: https://github.com/scailfin/flowserv-core/blob/master/LICENSE

.. image:: https://github.com/scailfin/flowserv-core/workflows/build/badge.svg
    :target: https://github.com/scailfin/flowserv-core/actions?query=workflow%3A%22build%22

.. image:: https://readthedocs.org/projects/flowserv-core/badge/?version=latest
    :target: https://flowserv-core.readthedocs.io/en/latest/?badge=latest
    :alt: Documentation Status

.. image:: https://codecov.io/gh/scailfin/flowserv-core/branch/master/graph/badge.svg
    :target: https://codecov.io/gh/scailfin/flowserv-core



.. figure:: https://github.com/scailfin/flowserv-core/blob/master/docs/figures/logo.png
    :align: center
    :alt: flowServ Logo



About
=====

This repository contains the implementation of the core infrastructure for the *Reproducible and Reusable Data Analysis Workflow Server* (**flowServ**). This is an experimental prototype to support reuse and evaluation of published data analysis pipelines as well as community benchmarks of data analysis algorithms. **flowServ** is not yet-another workflow engine. The aim instead is to provide a layer between a client (e.g. a Web user interface) and a workflow engine to facilitate the execution of a defined workflow templates (as shown in the figure below). *flowServ* is designed to be independent of the underlying workflow engine.

Workflow templates contain placeholders for workflow steps and/or input data and parameters that are provided by the user (e.g., by providing Docker containers that satisfy the workflow steps or uploading input data files). **flowServ** triggers and monitors the execution of the workflow for the given input values and maintains the workflow results. The API provides the functionality to submit new workflow runs and to retrieve the evaluation results of completed workflow runs.


.. figure:: https://github.com/scailfin/flowserv-core/blob/master/docs/figures/flowserv-overview.png
    :align: center
    :alt: ROB Architecture



**flowServ** was motivated by the `Reproducible Open Benchmarks for Data Analysis Platform (ROB) <https://github.com/scailfin/rob-ui>`_.  The goal of ROB is to allow user communities to evaluate the performance of their different data analysis algorithms in a controlled competition-style format. In ROB, the benchmark coordinator defines the workflow template along with input data. Benchmark participants provide their own implementation of the variable workfow steps. The workflow engine processes workflows on submission. Execution results are maintained by **flowServ** in an internal database. The goal of **flowServ** is to be a more generic platform that can not only be used for benchmarks but also for other types of data analysis workflows.



More Information
================

Workflow templates are motivated by the goal to allow users to run pre-defined data analytics workflows while providing their own input data, parameters, as well as their own code modules. Workflow templates are inspired by, but not limited to, workflow specifications for the `Reproducible Research Data Analysis Platform (REANA) <http://www.reanahub.io/>`_. The `Workflow Templates Section <https://github.com/scailfin/flowserv-core/blob/master/docs/workflow.rst>`_ provides further information about templates and their syntax. These templates are used by **flowServ** to run workflows and to maintain benchmark results.

The **flowServ** API defines the main interface to programmatically interact with the underlying database and workflow engine. The API implementation that is included in this repository provides a default serialization of all API resources as Python dictionaries. The API is intended to be used by Web applications. These applications can be build using different frameworks. The `current default Web API implementation for ROB <https://github.com/scailfin/rob-webapi-flask>`_ uses the `Flask web framework <https://flask.palletsprojects.com>`_.

ROB currently provides two different interfaces to interact with a Web API: the `Command Line Client <https://github.com/scailfin/rob-client>`_ and the `Web User Interface <https://github.com/scailfin/rob-ui>`_. See the respective repositories for further information on how to install and use these interfaces.

For an overview of ROB there are `slides <https://github.com/scailfin/presentations/blob/master/slides/ROB-Demo-MSDSE2019.pdf>`_ from the ROB Demo at the `Moore-Sloan Data Science Environment's annual summit 2019 <https://sites.google.com/msdse.org/summit2019/home>`_ and our `presentation <https://indico.cern.ch/event/822074/contributions/3471463/attachments/1865533/3067815/Reproducible_Benchmarks_for_Data_Analysis-v3.pdf>`_ at the `Analysis Systems Topical Workshop <https://indico.cern.ch/event/822074/>`_.

The full documentation is also available on `readthedocs.io <https://flowserv-core.readthedocs.io/en/latest/>`_.

Note
====

**flowServ** originated from the Reproducible Open Benchmarks for Data Analysis Platform (ROB). This repository replaces `Workflow Templates <https://github.com/scailfin/benchmark-templates>`_ and the `Reproducible Benchmark Engine <https://github.com/scailfin/benchmark-engine>`_ from an earlier version of ROB.


CHANGELOG.md
# Reproducible and Reusable Data Analysis Workflow Server - Changelog

### 0.1.0 - 2020-02-14

* Initial Version (from rob-core).


### 0.1.1 - 2020-02-20

* Add generic controller for remote workflow engines.


### 0.1.2 - 2020-02-23

* Read template metadata from project description file when adding a workflow template to the repository.
* Create default names for templates with missing or non-unique names.
* Include Short cuts to default repositories when creating workflow templates using the command-line interface.


### 0.2.0 - 2020-07-30

* Use SQLAlchemy as ORM for the database model.
* Run workflow templates from command-line for test purposes.
* Add methods for running and testing workflows in Python scripts (or Jupyter Notebooks).
* Include support and a command line interface for single-workflow applications.
* Install workflow templates from global repository.


### 0.2.1 - 2020-08-01

* Fix bug in created_at timestamp for workflow runs.
* Command-line interface to register new users.


### 0.3.0 - 2020-08-25

* Option to add specification of output file properties for display purposes in workflow specification.
* Rename data type element in parameter declarations and result schema columns to 'dtype'.
* Add manifest file option as parameter for test workflow runs in Jupyter.
* Test workflow runs using Docker engine.


### 0.3.1 - 2020-08-27

* Commit changes after run state update (issue \#48).


### 0.3.2 - 2020-08-28

* Allow Conditional Parameter Replacements (\#50).


### 0.3.3 - 2020-08-28

* Handle optional manifest element in workflow repository entries (\#52).


### 0.4.0 - 2020-09-16

* Refactor code to include abstract file store for alternative storage backends (\#54).
* Implement file store for S3 buckets.
* Avoid redundant storage of static files for workflow runs (\#55).
* Remove old workflow runs from database (\#9).


### 0.5.0 - 2020-10-01

* Allow user-defined workflow and group identifier (\#59).
* Merge `flowserv` and `flowapp` console scripts.
* User-defined keys for output files (\#60).
* Add functionality for asynchronous runs to flowserv application object.
* Change flowserv application run result object (\#61).
* Configure authentication policy via environment variables.
* Create separate groups for each workflow run in the app (\#62).
* RunResults `get_file` returns a single file object (\#63).
* Flowserv class for running workflows in notebook environments (\#64, \#65).


### 0.5.1 - 2020-10-03

* Create directory for SQLite database files (\#68).
* Type tests for enum and numeric parameters.
* Add serialized file handle to file listing in the application run result.
* Fix bug in run handle serialization for output files.
* Fix bug in Docker engine (return rundir in error case).
* Add ignore postproc option to `flowserv install` (\#69).
* File object for files that are uploaded as part of Flask requests.
* Make database web_app flag configurable using environment variable (\#70).


### 0.6.0 - 2020-11-17

* Add parameter declarations for lists and records (\#73).
* Rename run argument serialization helper functions.


### 0.7.0 - 2021-01-20

* Introduce `flowserv.service.api.APIFactory` to support multiple clients with different configuration on the same machine.
* Add functionality for user groups and authentication via the API client.
* Better support for accessing run result files via the API client.
* Include full functionality from [rob-client](https://github.com/scailfin/rob-client) into the `flowserv` CLI (\#72).
* Provide support for local and remote API's.
* Remove some of the obsolete code for running workflows in Jupyter Notebooks (now included in the new client interface `flowserv.client.Flowserv`).
* Drop support for Python 3.6.


### 0.7.1 - 2021-01-29

* Add option for passing an access token in API factory call.
* Ensure not to override *FLOWSERV_ASYNC* in `ClientAPI`.
* Add CLI environment context to support entry points for `flowserv` and `rob`.
* Extend serialized objects to contain additional resources (i.e., groups and runs) for authenticated users.


### 0.7.2 - 2021-02-03

* Change the location of the default data directory to be the OS-specific user cache directory.


### 0.8.0 - 2021-03-01

* Remove additional sub-directory `.flowserv` from default data directory path.
* New workflow template parameter type `actor` to support complete workflow steps as argument values.
* Serial engine workers to support execution of individual workflow steps using different backends.
* Remove separate Docker workflow controller


### 0.8.1 - 2021-05-03

* Fix issue when creating database object on Windows.
* Fix issue with schema validator on Windows.
* Fix issue with bucket store unit tests on Windows.
* Fix unit test issues for Windows.
* Fix issue with subprocess worker on Windows.


### 0.9.0 - 2021-06-15

* Make `BucketStore` generic for storing workflow files in different types of cloud storage buckets (\#80).
* Add Google Cloud File Storage bucket implementation (\#80).
* Support different storage volumes for different types of workers.
* Add support for workflow steps that are implemented as Jupyter Notebooks (\#79).
* Include flowapp to run workflow templates using streamlit GUI.


### 0.9.1 - 2021-06-22

* Add option to resolve references to external files when reading Yaml files.
* Support inclusion and import of files in instructions markdown text.


### 0.9.2 - 2021-09-30

* Reset hard-coded workflow repository URL to point to the master branch.


### 0.9.3 - 2022-05-24

* Fix issue with `S3Volume.walk` return value (\#95)


### 0.9.4 - 2022-05-25

* Make `path` argument for `copy_files` less restrictive (\#96)


flowserv/config.py
# This file is part of the Reproducible and Reusable Data Analysis Workflow
# Server (flowServ).
#
# Copyright (C) 2019-2021 NYU.
#
# flowServ is free software; you can redistribute it and/or modify it under the
# terms of the MIT License; see LICENSE file for more details.

"""Collection of configuration parameters for different components of flowserv.

The module contains helper classes to get configuration values from environment
variables and to customize the configuration settings.
"""

from __future__ import annotations
from appdirs import user_cache_dir
from typing import Any, Dict, Optional, Union

import os

from flowserv.volume.fs import FStore

import flowserv.error as err
import flowserv.util as util

# --
# -- Environment variables and default values
# --

# -- API ----------------------------------------------------------------------

"""Names of environment variables that are used to configure the API."""
# Base directory to store uploaded files and submission results
FLOWSERV_BASEDIR = 'FLOWSERV_API_DIR'
# Host name for API server
FLOWSERV_API_HOST = 'FLOWSERV_API_HOST'
# Name of the API instance
FLOWSERV_API_NAME = 'FLOWSERV_API_NAME'
# API application path
FLOWSERV_API_PATH = 'FLOWSERV_API_PATH'
# API server port on host
FLOWSERV_API_PORT = 'FLOWSERV_API_PORT'
# Url protocol prefix
FLOWSERV_API_PROTOCOL = 'FLOWSERV_API_PROTOCOL'


"""Default values for environment variables."""
DEFAULT_HOST = 'localhost'
DEFAULT_NAME = 'Reproducible and Reusable Data Analysis Workflow Server (API)'
DEFAULT_PATH = '/flowserv/api/v1'
DEFAULT_PORT = 5000
DEFAULT_PROTOCOL = 'http'


def API_DEFAULTDIR() -> str:
    """The default API base directory is a subfolder in the users data cache
    directory.

    The default user cache directory is different for different OS's. We use
    appdirs.user_cache_dir to get the directory.

    Returns
    -------
    string
    """
    return user_cache_dir(appname=__name__.split('.')[0])


def API_URL(env: Dict) -> str:
    """Get the base URL for the API from the respective environment variables
    'FLOWSERV_API_HOST', 'FLOWSERV_API_PATH', and 'FLOWSERV_API_PORT' in the
    given configuration dictionary.

    Parameters
    ----------
    env: dict
        Configuration object that provides access to configuration
        parameters in the environment.

    Returns
    -------
    string
    """
    protocol = env.get(FLOWSERV_API_PROTOCOL, DEFAULT_PROTOCOL)
    host = env.get(FLOWSERV_API_HOST, DEFAULT_HOST)
    port = to_int(env.get(FLOWSERV_API_PORT, DEFAULT_PORT))
    if port != 80:
        host = '{}:{}'.format(host, port)
    path = env.get(FLOWSERV_API_PATH, DEFAULT_PATH)
    if not path.startswith('/'):
        path = '/' + path
    return '{}://{}{}'.format(protocol, host, path)


# -- Application --------------------------------------------------------------

"""Names of environment variables that configure the application."""

FLOWSERV_APP = 'FLOWSERV_APP'
FLOWSERV_GROUP = 'FLOWSERV_GROUP'

# Respective variable names for ROB.
ROB_BENCHMARK = 'ROB_BENCHMARK'
ROB_SUBMISSION = 'ROB_SUBMISSION'


def APP() -> str:
    """Get the value for the FLOWSERV_APP variable from the environment. Raises
    a missing configuration error if the value is not set.
    Returns
    -------
    string
    """
    app_key = os.environ.get(FLOWSERV_APP)
    if not app_key:
        raise err.MissingConfigurationError('workflow identifier')
    return app_key


# -- Auth ---------------------------------------------------------------------

"""Names of environment variables that are used to configure the authentication
module.
"""
# Access token for the command line interface
FLOWSERV_ACCESS_TOKEN = 'FLOWSERV_ACCESS_TOKEN'
# Time period for which an API key is valid
FLOWSERV_AUTH_LOGINTTL = 'FLOWSERV_AUTH_TTL'
# Authentication policy
FLOWSERV_AUTH = 'FLOWSERV_AUTH'


"""Default values for environment variables."""
DEFAULT_LOGINTTL = 24 * 60 * 60
# Access policies
AUTH_DEFAULT = 'default'
AUTH_OPEN = 'open'

"""Default user."""
DEFAULT_USER = '0' * 8


# -- Backend ------------------------------------------------------------------

"""Environment variables that control the configuration of the workflow
controllers.
"""
# Name of the class that implements the workflow controller interface
FLOWSERV_BACKEND_CLASS = 'FLOWSERV_BACKEND_CLASS'
# Name of the module that contains the workflow controller implementation
FLOWSERV_BACKEND_MODULE = 'FLOWSERV_BACKEND_MODULE'
# Flag indicating whether workflows are executed asynchronously or blocking.
FLOWSERV_ASYNC = 'FLOWSERV_ASYNCENGINE'
DEFAULT_ASYNC = 'True'

# Poll interval
FLOWSERV_POLL_INTERVAL = 'FLOWSERV_POLLINTERVAL'
DEFAULT_POLL_INTERVAL = 2


# -- Client -------------------------------------------------------------------

"""Environment variables for the command line interface."""
# Define the type of client that the command line interface uses.
FLOWSERV_CLIENT = 'FLOWSERV_CLIENT'
LOCAL_CLIENT = 'local'
REMOTE_CLIENT = 'remote'


# -- Database -----------------------------------------------------------------

"""Environment variable that contains the database connection string."""
FLOWSERV_DB = 'FLOWSERV_DATABASE'
FLOWSERV_WEBAPP = 'FLOWSERV_WEBAPP'


# -- File store ---------------------------------------------------------------

"""Environment variables that are used to configure the file store that is used
to maintain files for workflow templates, user group uploads, and workflow runs.
"""
# Identifier of the file store class.
FLOWSERV_FILESTORE = 'FLOWSERV_FILESTORE'


# --
# -- Configuration settings
# --


class Config(dict):
    """Helper class that extends a dictionary with dedicated methods to set
    individual parameters in the flowserv configuration.

    Methods with lower case names are setters for configuration parameters. All
    setters return a reference to the configuration object itself to allows
    chanining the setter calls.
    """
    def __init__(self, defaults: Optional[Dict] = None):
        """Initialize the dictionary.

        Parameters
        ----------
        defaults: dict, default=None
            Dictionary with default settings.
        """
        if defaults is not None:
            super(Config, self).__init__(**defaults)

    def auth(self) -> Config:
        """Set the authentication method to the default value that requires
        authentication.

        Returns
        -------
        flowserv.config.Config
        """
        self[FLOWSERV_AUTH] = AUTH_DEFAULT
        return self

    def basedir(self, path: str) -> Config:
        """Set the flowserv base directory.

        Parameters
        ----------
        path: string
            Path to the base directory for all workflow files.

        Returns
        -------
        flowserv.config.Config
        """
        self[FLOWSERV_BASEDIR] = os.path.abspath(path)
        return self

    def database(self, url: str) -> Config:
        """Set the database connect Url.

        Parameters
        ----------
        url: string
            Database connect Url.

        Returns
        -------
        flowserv.config.Config
        """
        self[FLOWSERV_DB] = url
        return self

    def multiprocess_engine(self) -> Config:
        """Set configuration to use the serial multi-porcess workflow controller
        as the default backend.

        Returns
        -------
        flowserv.config.Config
        """
        self[FLOWSERV_BACKEND_MODULE] = 'flowserv.controller.serial.engine.base'
        self[FLOWSERV_BACKEND_CLASS] = 'SerialWorkflowEngine'
        return self

    def open_access(self) -> Config:
        """Set the authentication method to open access.

        Returns
        -------
        flowserv.config.Config
        """
        self[FLOWSERV_AUTH] = AUTH_OPEN
        return self

    def run_async(self) -> Config:
        """Set the run asynchronous flag to True.

        Returns
        -------
        flowserv.config.Config
        """
        self[FLOWSERV_ASYNC] = True
        return self

    def run_sync(self) -> Config:
        """Set the run asynchronous flag to False.

        Returns
        -------
        flowserv.config.Config
        """
        self[FLOWSERV_ASYNC] = False
        return self

    def token_timeout(self, timeout: int) -> Config:
        """Set the authentication token timeout interval.

        Returns
        -------
        flowserv.config.Config
        """
        self[FLOWSERV_AUTH_LOGINTTL] = timeout
        return self

    def volume(self, config: dict) -> Config:
        """Set configuration object for the file storage volume.

        Parameters
        ----------
        config: dict
            Volume configuration information for the storage volume factory.

        Returns
        -------
        flowserv.config.Config
        """
        self[FLOWSERV_FILESTORE] = config
        return self

    def webapp(self) -> Config:
        """Set the web app flag to True.

        Returns
        -------
        flowserv.config.Config
        """
        self[FLOWSERV_WEBAPP] = True
        return self


# -- Initialize configuration from environment variables ----------------------

def to_bool(value: Any) -> bool:
    """Convert given value to Boolean. Only of the string matches 'True' (ignoring
    case) the result will be True.

    Parameters
    ----------
    value: any
        Expects a string representation of a Boolean value.

    Returns
    -------
    bool
    """
    try:
        return True if value.lower() == 'true' else False
    except AttributeError:
        return False


def to_float(value: Any) -> float:
    """Convert given value to float.

    Parameters
    ----------
    value: any
        Expects an integer, float, or a string representation of an integer
        or float value.

    Returns
    -------
    int
    """
    try:
        return float(value)
    except ValueError:
        return None


def to_int(value: Any) -> int:
    """Convert given value to integer.

    Parameters
    ----------
    value: any
        Expects an integer value or a string representation of an integer value.

    Returns
    -------
    int
    """
    try:
        return int(value)
    except ValueError:
        return None


def read_config_obj(filename: Union[str, Dict]) -> Dict:
    """Read configuration object from a file.

    This function only attempts to read an object from disk if the type of the
    filename argument is string.

    Parameters
    ----------
    filename: str or dict
        Path to file on disk.

    Returns
    -------
    dict
    """
    if isinstance(filename, dict):
        return filename
    return util.read_object(filename=filename)


"""List of environment variables and their default settings and an optional
value case function.
"""

ENV = [
    (FLOWSERV_BASEDIR, API_DEFAULTDIR(), None),
    (FLOWSERV_API_HOST, DEFAULT_HOST, None),
    (FLOWSERV_API_NAME, DEFAULT_NAME, None),
    (FLOWSERV_API_PATH, DEFAULT_PATH, None),
    (FLOWSERV_API_PORT, DEFAULT_PORT, to_int),
    (FLOWSERV_API_PROTOCOL, DEFAULT_PROTOCOL, None),
    (FLOWSERV_APP, None, None),
    (FLOWSERV_AUTH_LOGINTTL, DEFAULT_LOGINTTL, to_int),
    (FLOWSERV_AUTH, AUTH_DEFAULT, None),
    (FLOWSERV_BACKEND_CLASS, None, None),
    (FLOWSERV_BACKEND_MODULE, None, None),
    (FLOWSERV_POLL_INTERVAL, DEFAULT_POLL_INTERVAL, to_float),
    (FLOWSERV_ACCESS_TOKEN, None, None),
    (FLOWSERV_CLIENT, LOCAL_CLIENT, None),
    (FLOWSERV_DB, None, None),
    (FLOWSERV_WEBAPP, 'False', to_bool),
    (FLOWSERV_FILESTORE, FStore(basedir=API_DEFAULTDIR()), read_config_obj)
]


def env() -> Config:
    """Get configuration parameters from the environment."""
    config = Config()
    for var, default, cast in ENV:
        value = os.environ.get(var, default)
        if value is not None:
            if cast is not None:
                value = cast(value)
            config[var] = value
    return config
